Derivation crashes for case object

## minimized code

```Scala
trait Tc[A]
object Tc {
  inline def derived[A](given A: deriving.Mirror.Of[A]): Tc[A] = new Tc[A] {}
}

case object Thing derives Tc
```


<details>
<summary>Stack trace</summary>

```scala
Exception in thread "main" java.lang.AssertionError: assertion failed
        at dotty.DottyPredef$.assertFail(DottyPredef.scala:16)
        at dotty.tools.dotc.typer.Implicits.companionPath(Implicits.scala:912)
        at dotty.tools.dotc.typer.Implicits.mirrorFor$2(Implicits.scala:981)
        at dotty.tools.dotc.typer.Implicits.synthesizedProductMirror$$anonfun$1$$anonfun$1(Implicits.scala:991)
        at dotty.tools.dotc.typer.Implicits.synthesizedMirror$$anonfun$1$$anonfun$1(Implicits.scala:1080)
        at dotty.tools.dotc.typer.Implicits.trySpecialCases$1(Implicits.scala:1123)
        at dotty.tools.dotc.typer.Implicits.trySpecialCases$2$$anonfun$1(Implicits.scala:1125)
        at dotty.tools.dotc.ast.Trees$Tree.orElse(Trees.scala:185)
        at dotty.tools.dotc.typer.Implicits.trySpecialCases$1(Implicits.scala:1125)
        at dotty.tools.dotc.typer.Implicits.trySpecialCases$2$$anonfun$1(Implicits.scala:1125)
        at dotty.tools.dotc.ast.Trees$Tree.orElse(Trees.scala:185)
        at dotty.tools.dotc.typer.Implicits.trySpecialCases$1(Implicits.scala:1125)
        at dotty.tools.dotc.typer.Implicits.trySpecialCases$2$$anonfun$1(Implicits.scala:1125)
        at dotty.tools.dotc.ast.Trees$Tree.orElse(Trees.scala:185)
        at dotty.tools.dotc.typer.Implicits.trySpecialCases$1(Implicits.scala:1125)
        at dotty.tools.dotc.typer.Implicits.trySpecialCases$2$$anonfun$1(Implicits.scala:1125)
        at dotty.tools.dotc.ast.Trees$Tree.orElse(Trees.scala:185)
        at dotty.tools.dotc.typer.Implicits.trySpecialCases$1(Implicits.scala:1125)
        at dotty.tools.dotc.typer.Implicits.trySpecialCases$2$$anonfun$1(Implicits.scala:1125)
        at dotty.tools.dotc.ast.Trees$Tree.orElse(Trees.scala:185)
        at dotty.tools.dotc.typer.Implicits.trySpecialCases$1(Implicits.scala:1125)
        at dotty.tools.dotc.typer.Implicits.trySpecialCases$2$$anonfun$1(Implicits.scala:1125)
        at dotty.tools.dotc.ast.Trees$Tree.orElse(Trees.scala:185)
        at dotty.tools.dotc.typer.Implicits.trySpecialCases$1(Implicits.scala:1125)
        at dotty.tools.dotc.typer.Implicits.trySpecialCases$2$$anonfun$1(Implicits.scala:1125)
        at dotty.tools.dotc.ast.Trees$Tree.orElse(Trees.scala:185)
        at dotty.tools.dotc.typer.Implicits.trySpecialCases$1(Implicits.scala:1125)
        at dotty.tools.dotc.typer.Implicits.inferImplicitArg(Implicits.scala:1130)
        at dotty.tools.dotc.typer.Typer.inferImplicitArg(Typer.scala:83)
        at dotty.tools.dotc.typer.Typer.implicitArgs$2(Typer.scala:2697)
        at dotty.tools.dotc.typer.Typer.addImplicitArgs$4(Typer.scala:2720)
        at dotty.tools.dotc.typer.Typer.adaptNoArgsImplicitMethod$2(Typer.scala:2787)
        at dotty.tools.dotc.typer.Typer.adaptNoArgs$1(Typer.scala:2957)
        at dotty.tools.dotc.typer.Typer.adapt1(Typer.scala:3185)
        at dotty.tools.dotc.typer.Typer.op$3(Typer.scala:2578)
        at dotty.tools.dotc.typer.Typer.adapt(Typer.scala:2579)
        at dotty.tools.dotc.typer.Typer.readapt$1(Typer.scala:2588)
        at dotty.tools.dotc.typer.Typer.adapt1(Typer.scala:3169)
        at dotty.tools.dotc.typer.Typer.op$3(Typer.scala:2578)
        at dotty.tools.dotc.typer.Typer.adapt(Typer.scala:2579)
        at dotty.tools.dotc.typer.Typer.op$1(Typer.scala:2247)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2256)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2259)
        at dotty.tools.dotc.typer.Deriving$Deriver.typeclassInstance$1$$anonfun$1$$anonfun$1(Deriving.scala:291)
        at dotty.tools.dotc.ast.tpd$.polyDefDef(tpd.scala:262)
        at dotty.tools.dotc.typer.Deriving$Deriver.syntheticDef$1(Deriving.scala:295)
        at dotty.tools.dotc.typer.Deriving$Deriver.syntheticDefs$$anonfun$1(Deriving.scala:297)
        at scala.collection.StrictOptimizedIterableOps.map(StrictOptimizedIterableOps.scala:100)
        at scala.collection.StrictOptimizedIterableOps.map$(StrictOptimizedIterableOps.scala:87)
        at scala.collection.mutable.ListBuffer.map(ListBuffer.scala:38)
        at dotty.tools.dotc.typer.Deriving$Deriver.syntheticDefs(Deriving.scala:297)
        at dotty.tools.dotc.typer.Deriving$Deriver.finalize(Deriving.scala:302)
        at dotty.tools.dotc.typer.Typer.finalize$2(Typer.scala:2319)
        at dotty.tools.dotc.typer.Typer.$anonfun$45(Typer.scala:2326)
        at scala.collection.immutable.List.mapConserve(List.scala:444)
        at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2326)
        at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:1812)
        at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2141)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2208)
        at dotty.tools.dotc.typer.Typer.op$1(Typer.scala:2247)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2256)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2259)
        at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2281)
        at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2325)
        at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:1938)
        at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2182)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2209)
        at dotty.tools.dotc.typer.Typer.op$1(Typer.scala:2247)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2256)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2259)
        at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2367)
        at dotty.tools.dotc.typer.FrontEnd.liftedTree1$2(FrontEnd.scala:78)
        at dotty.tools.dotc.typer.FrontEnd.typeCheck$$anonfun$1(FrontEnd.scala:83)
        at dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
        at dotty.tools.dotc.typer.FrontEnd.monitor(FrontEnd.scala:42)
        at dotty.tools.dotc.typer.FrontEnd.typeCheck(FrontEnd.scala:84)
        at dotty.tools.repl.REPLFrontEnd.runOn(ReplFrontEnd.scala:23)
        at dotty.tools.dotc.Run.op$4(Run.scala:167)
        at dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:177)
        at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
        at dotty.tools.dotc.Run.runPhases$2(Run.scala:177)
        at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:185)
        at dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
        at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:68)
        at dotty.tools.dotc.Run.compileUnits(Run.scala:192)
        at dotty.tools.dotc.Run.compileUnits(Run.scala:134)
        at dotty.tools.repl.ReplCompiler.runCompilationUnit(ReplCompiler.scala:156)
        at dotty.tools.repl.ReplCompiler.compile(ReplCompiler.scala:166)
        at dotty.tools.repl.ReplDriver.compile(ReplDriver.scala:229)
        at dotty.tools.repl.ReplDriver.interpret(ReplDriver.scala:193)
        at dotty.tools.repl.ReplDriver.loop$1(ReplDriver.scala:127)
        at dotty.tools.repl.ReplDriver.runUntilQuit$$anonfun$1(ReplDriver.scala:130)
        at dotty.tools.repl.ReplDriver.withRedirectedOutput(ReplDriver.scala:148)
        at dotty.tools.repl.ReplDriver.runUntilQuit(ReplDriver.scala:130)
        at dotty.tools.repl.Main$.main(Main.scala:6)
        at dotty.tools.repl.Main.main(Main.scala)
```
</details>

--- a/compiler/src/dotty/tools/dotc/typer/Implicits.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Implicits.scala
@@ -909,7 +909,7 @@ trait Implicits { self: Typer =>
   /** A path referencing the companion of class type `clsType` */
   private def companionPath(clsType: Type, span: Span)(implicit ctx: Context) = {
     val ref = pathFor(clsType.companionRef)
-    assert(ref.symbol.is(Module) && ref.symbol.companionClass == clsType.classSymbol)
+    assert(ref.symbol.is(Module) && (clsType.classSymbol.is(ModuleClass) || (ref.symbol.companionClass == clsType.classSymbol)))
     ref.withSpan(span)
   }
 

--- a/tests/run/typeclass-derivation3.scala
+++ b/tests/run/typeclass-derivation3.scala
@@ -19,6 +19,9 @@ object datatypes {
   sealed trait Either[+L, +R] extends Product with Serializable derives Eq, Pickler, Show
   case class Left[L](x: L) extends Either[L, Nothing]
   case class Right[R](x: R) extends Either[Nothing, R]
+
+  // a case object
+  case object Thing derives Eq, Pickler, Show
 }
 
 object typeclasses {

